tufaceous-lib/src/artifact.rs

```diff
@@ -58,4 +78,129 @@ impl AddArtifact {
     pub fn version(&self) -> &str {
         &self.version
     }
+
+    /// Returns the source for this artifact.
+    pub fn source(&self) -> &ArtifactSource {
+        &self.source
+    }
+
+    /// Writes this artifact to the specified writer.
+    pub(crate) fn write_to<W: Write>(&self, writer: &mut W) -> Result<()> {
+        match &self.source {
+            ArtifactSource::File(path) => {
+                let mut reader = File::open(path)?;
+                std::io::copy(&mut reader, writer)?;
+            }
+            ArtifactSource::Memory(buf_list) => {
+                for chunk in buf_list {
+                    writer.write_all(chunk)?;
+                }
+            }
+            ArtifactSource::Fake { size } => {
+                let known = self.kind.to_known();
+                if known == Some(KnownArtifactKind::Host)
+                    || known == Some(KnownArtifactKind::Trampoline)
+                {
+                    write_host_tarball_fake_artifact(*size, writer)?;
+                } else {
+                    write_generic_fake_artifact(*size, writer)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Writes a fake artifact with no internal structure.
+fn write_generic_fake_artifact<W: Write>(
+    size: u64,
+    writer: &mut W,
+) -> io::Result<()> {
+    // Don't try and get the size exactly right
+    let times = (size as usize) % FILLER_TEXT.len();
```

Comment:
After seeing `FillerReader` below, maybe we could push this iterator stuff down into it? (https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=05b4fe9f9c3dcf58ca440d114909b186) Then we could use `FillerReader` here too.

Reply:
Hmm, I've done this for the generic artifact (and added `BufWriter` around it), but the tarball builder needs a `Read` implementation sadly. It would be nice if there were a way to get a handle to write data to ([that's how the zip crate works](https://docs.rs/zip/0.6.4/zip/struct.ZipWriter.html#method.start_file)), but I didn't see anything like that in the tar builder API.

Reply:
I can't really think of a way to do this with the iterator that's less complicated.

Reply:
I added a comment explaining the `FillerReader` situation, including a link to https://github.com/alexcrichton/tar-rs/issues/304.

Reply:
I tried out the `std::iter` approach and sadly it was roughly 10x slower (at least in debug mode, which is what most people will use) than just writing 16 bytes at a time.

If we really want to get the size exactly right then we can do more math I think, but also FILLER_TEXT is just 16 bytes long so it doesn't really matter I think? For all practical purposes size will never be smaller than 16 bytes.